### PR TITLE
Allow to disable autoClose on Xpi

### PR DESCRIPTION
**Pull request by willdurand**
With the `keepAlive` option in `Xpi`, I am afraid that we can run into a race condition/timing issue if the OS closes the file descriptor after we checked `isOpen` and before we open a stream. I was not able to trigger this scenario but the code lets me think that it is possible.

I switched to a new approach that allows us to not auto-close the file descriptor. That way we can reuse the same file descriptor (once open) and `close()` it manually. I added a noop `close()` method on the public/common API because it is much easier to deal with.

I added real zip files to avoid mocking too many things.

I changed the `close` event to `end` otherwise `getFiles()` never completes when we disable `autoClose` (because it needs the `close` event to be emitted). I tried to reproduce the issue mentioned in https://github.com/mozilla/addons-linter/pull/43 but I was not able to do so... I tried to run test cases in parallel (up to 1000 concurrent promises) and many times (up to 10k synchronous runs), multiple times. I could not trigger a race condition. I believe this commit prevents the race condition to happen: https://github.com/thejoshwolfe/yauzl/commit/594cd7ea712ec21f8f6eb80d9ac52b9ad7d1b76b#diff-168726dbe96b3ce427e7fedce31bb0bcR215-R220

I cannot confirm though. I tried to reproduce a race condition with the old addons-validator/linter, before https://github.com/mozilla/addons-linter/pull/43. I could not trigger any race condition.

### CODE REVIEW

File: src/io/xpi.spec.ts [hunk 1]
```diff
@@ -94,12 +94,12 @@ describe(__filename, () => {
   });
 
   const createXpi = ({
+    autoClose = true,
```
Comment (kumar303) [inline src/io/xpi.spec.ts]: Is there any case for `autoClose: false`? What would need to do that?

File: src/io/xpi.spec.ts [hunk 2]
```diff
@@ -618,4 +662,45 @@ describe(__filename, () => {
...
+
+      expect(xpi.zipfile).not.toBeDefined();
+
+      // This is used to trigger a call to `open()` using the public API.
+      await xpi.getFiles();
+
+      // `zipfile` is created when `getFiles()` is called.
+      expect(xpi.zipfile).toBeDefined();
+      // @ts-ignore
+      expect(xpi.zipfile.isOpen).toEqual(true);
```
[the remark this answers is not in the document]
Comment (willdurand) [reply src/io/xpi.spec.ts]: No, but I can avoid the `@ts-ignore` comment by writing `xpi.zipfile && xpi.zipfile.isOpen`.

File: src/io/xpi.spec.ts [hunk 3]
```diff
@@ -618,4 +662,45 @@ describe(__filename, () => {
...
+
+    it('does nothing when autoClose is enabled', () => {
+      const xpi = new Xpi({
+        autoClose: true,
+        filepath: '',
+        stderr: createFakeStderr(),
+      });
+
+      // @ts-ignore
+      xpi.zipfile = { close: jest.fn() };
```
[the remark this answers is not in the document]
Comment (willdurand) [reply src/io/xpi.spec.ts]: Yep, I think I can try that. This test suite comes from addons-linter and I was not brave enough to fix everything :( I left some comments about that problem though.

File: src/io/xpi.spec.ts (hunk 1 above)
Comment (willdurand) [reply src/io/xpi.spec.ts]: mmh, for `autoClose: false` or `true`? `autoClose: false` means we can manually close the file after a bunch of read operations, which is a great improvement.

**Pull request merged**